backend/cloud_inquisitor/plugins/types/accounts.py

```diff
@@ -0,0 +1,443 @@
+import logging
+from abc import ABC, abstractmethod
+from datetime import datetime
+
+from cloud_inquisitor.constants import PLUGIN_NAMESPACES
+from sqlalchemy import and_, or_, desc
+from sqlalchemy.exc import SQLAlchemyError
+from sqlalchemy.orm import aliased
+
+from cloud_inquisitor import get_plugin_by_name
+from cloud_inquisitor.database import db
+from cloud_inquisitor.exceptions import AccountException, InquisitorError
+from cloud_inquisitor.schema import AccountProperty, Account, AccountType
+from cloud_inquisitor.utils import to_camelcase
+
+
+class BaseAccount(ABC):
+    class_properties = []
+
+    """Base type object for Account objects"""
+    def __init__(self, account):
+        self.account = account
+        self.log = logging.getLogger(self.__class__.__module__)
+
+    def __getattr__(self, item):
+        return self.get_property(item)
+
+    def __str__(self):
+        return self.account.account_name
+
+    def get_property(self, item):
+        for prop in self.account.properties:
+            if prop.name == item:
+                return prop
+
+        raise AttributeError(item)
+
+    def set_property(self, name, value):
+        if type(value) == datetime:
+            value = value.isoformat()
+        else:
+            value = value
+
+        try:
+            prop = self.get_property(name)
+            if prop.value == value:
+                return False
+
+            prop.value = value
+
+        except AttributeError:
+            prop = AccountProperty()
+            prop.account_id = self.id
+            prop.name = name
+            prop.value = value
+
+        db.session.add(prop)
+
+        return True
+
+    def delete_property(self, name):
+        try:
+            self.log.debug('Removing property {} from {}'.format(name, self.id))
+            prop = getattr(self, name)
+            db.session.delete(prop)
+            self.properties.remove(prop)
+
+            return True
+        except AttributeError:
+            return False
+
+    def save(self, *, auto_commit=True):
+        try:
+            db.session.add(self.account)
+            if auto_commit:
+                db.session.commit()
+        except SQLAlchemyError as ex:
+            self.log.exception('Failed updating account: {}'.format(ex))
+            db.session.rollback()
+
+    def delete(self, *, auto_commit=True):
+        try:
+            db.session.delete(self.account)
+            if auto_commit:
+                db.session.commit()
+        except SQLAlchemyError:
+            self.log.exception('Failed deleting account: {}'.format(self.id))
+            db.session.rollback()
+
+    def to_json(self, is_admin=False):
```

Comment:
This function name seems misleading.  It's called `to_json` but it returns a dict instead of a json string.

Reply:
I don't disagree, but this is a pattern all for all types. We can look at refactoring this later, but its out of scope for this particular refactor.

Since we use a dynamic introspection pattern to convert objects to json encodable output we cannot rename the function in a single place, we'd have to do it for everything in one go.

Reply:
Okay thank you for the explanation.  I'll create a ticket.

Reply:
https://github.com/RiotGames/cloud-inquisitor/issues/174